[31935] WIP: Enterprise activation frontend

ingamai opened the pull request:
### ToDo

- [x] New frontend for enterprise page (Text, buttons, info boxes)
  - [x] Adapt homescreen block to use the same strings 
- [x] **Happy Path**: A click on "Start free trial" should open a modal, with multiple steps
  - [x] (1) Collect necessary user data
  - [x] (2) Waiting screen while the email is not confirmed
  - [x] (3) Show enterprise onboarding video
- [x] Catch cases where users deviate from the Happy path:
  - [x] The user closes the modal before confirming the mail: When going through the steps in the modal, change background dynamically to the current state. So when the modal is closed, the current state is still visible.
  - [x] The user reloads the page before confirming the mail: The data should still be visible on the page. Implemented via the Augur in key stored in the database. With that we can collect the data after reload from the API again.
  - [x] When the user waited too long, to confirm the mail (currently more than 5 minutes), he can resend the confirmation mail via the link and thus still confirm.
  - [x] When I accidentally wrote a wrong email address, I can try with another mail. (Works basically, but the modal still shows the old email address).
- [x] When I have a new token, I want to paste it in the according area and the information should update accordingly.
- [x] A second admin should be able to delete the token of the first admin
- [x] Make the used Augur API configurable
- [x] Change texts
- [x] Change edge Augur URL to correct URL
- [x] '(expired)' status behind user name? **This is due to augur-edge returning an expired license for testing**
- [x] Test cases


### Out of Scope

* Pre-fill form when clicking on "try with another mail"
* Add Company and Domain to the info screen (currently only name and email are shown) --> Extend Token structure
* Add information of when we send the confirmation email

ingamai committed:
Build frontend for new admin enterprise page

ingamai committed:
Successful happy path; WIP: error handling

ingamai committed:
Resend link working; form styles adapted; notifications added

ingamai committed:
Info boxes template added as .sass and .lsg

ingamai committed:
Links and images added to enterprise info boxes

ingamai committed:
Add quote link and onboarding video

ingamai committed:
Add limit for retrying to confirm mail

machisuji commented on app/views/enterprises/_info.html.erb:
```diff
@@ -1,28 +1,114 @@
+See docs/COPYRIGHT.rdoc for more details.
+
+++#%>
+<% content_for :header_tags do %>
```
@ingamai I found the problem. This doesn't work in partials (_info.html.erb). If you move this bit to show.html.erb t should work!

HDinger commented:
Tickets:
https://community.openproject.com/projects/saas/work_packages/32422/activity
https://community.openproject.com/projects/275/work_packages/32892/activity
https://community.openproject.com/projects/openproject/work_packages/32894/activity

ingamai committed:
Adapt background page content to state too

ingamai committed:
cleanup and DRY; youtube video added

HDinger commented on config/locales/js-en.yml:
```diff
@@ -181,6 +181,50 @@ en:
+          label_starts_at: "Starts at"
+          label_expires_at: "Expires at"
+          label_status: "Status:"
+          mail_not_received: "You did not receive a mail? Resend mail or"
```
Please take care that you use `mail` and `email` consistently. In Englisch, we use `email`. Also please have a look at upper and lower case at the beginning of a string.

HDinger commented on config/locales/js-en.yml:
```diff
@@ -181,6 +181,50 @@ en:
+          resend_warning: "Could not resend mail"
+          session_timeout: "Your session timed out. Please try to resend mail."
+          status_waiting: "email sent - waiting for confirmation"
+          status_confirmed: "confirmed ✓"
```
This is just a guess, but I can imagine that the checkmark could lead to issues when translating the string. Could you please check whether we could display this in any other way. Maybe with an icon?

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial-form/ee-trial-form.component.ts:
```diff
@@ -0,0 +1,80 @@
+  // displays message for user
+  public checkMailField() {
+    if (this.trialForm.value.email !== '' && this.trialForm.controls.email.errors) {
+      this.eeTrialService.errorMsg = 'Invalid e-mail address';
```
Please localise this.

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial.modal.html:
```diff
@@ -0,0 +1,77 @@
+       data-indicator-name="modal"
+       tabindex="0">
+    <div class="op-modal--modal-header">
+      <h3 *ngIf="!eeTrialService.status"> {{ text.heading_test_ee }}</h3>
```
Instead of using three elements, just show one and change the text according to the status in the .ts file. 
`<h3 [textContent]="headerText()"> </h3>`

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial.modal.html:
```diff
@@ -0,0 +1,77 @@
+                    height="400"
+                    width="100%"
+                    allowfullscreen=""
+                    src="https://www.youtube.com/embed/zLMSydhFSkw?autoplay=1">
```
You can extract the URL into a constant so that is easier to find later when the URL should change.

HDinger commented on app/assets/stylesheets/content/_info_boxes.lsg:
```diff
@@ -0,0 +1,73 @@
+# Info boxes
```
👍 Thanks for adding it to the LSG

HDinger commented on app/assets/stylesheets/content/_info_boxes.sass:
```diff
@@ -0,0 +1,63 @@
+    margin: 15px auto
+
+  &.-centered
+    max-width: 1140px
```
Why exactly this number (same for the base class)? And why is `-centered` wider than the base class?

ingamai replied:
@HDinger This is a very good question 😅 I will only give the base class a max-width (because I found it quite strange looking on a big screen if the elements span the whole width). I  used 1140px because this is the max-width of the content on openproject.org, on which I based the template for the info boxes. 
Should I change that to another value?

HDinger commented on app/assets/stylesheets/content/_enterprise.sass:
```diff
@@ -14,3 +14,8 @@
+.upsale--actions
+  display: flex
+  flex-wrap: wrap
+  align-items: baseline
```
In many files the last (empty) line is missing. Pease add this for code climate reasons. 🙈

HDinger commented on app/views/enterprises/_info.html.erb:
```diff
@@ -1,28 +1,121 @@
+     data-cb-plan-id="enterprise-edition---annual-user-license">
+    <%= t('admin.enterprise.book_now') %>
+  </a>
+  <a href="https://www.openproject.org/upgrade-enterprise-edition/" target="_blank">
```
Please extract these links as constants and collect them at one place. Thus it will be much easier to replace them in case the URL changes.

HDinger commented on app/views/enterprises/_info.html.erb:
```diff
@@ -1,28 +1,121 @@
+  <h2><%= t('homescreen.blocks.upsale.description') %></h2>
+  <div class="info-boxes--container">
+    <div class="info-boxes--item">
+      <%= image_tag "https://1t1rycb9er64f1pgy2iuseow-wpengine.netdna-ssl.com/wp-content/uploads/2017/06/icon-security-features.png",
```
Is there a reason to use the cryptical URL instead of adding the pictures to our image folder?

ingamai replied:
@HDinger I added the pictures to the images folder.

HDinger commented on frontend/src/app/components/enterprise/enterprise-base.component.html:
```diff
@@ -0,0 +1,17 @@
+  {{ text.you_contribute }}
+</p>
+
+<div *ngIf="eeTrialService.status === 'mailSubmitted'">
```
The check for the status is used quite often. I think it makes sense to extract it to a method.

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial.modal.sass:
```diff
@@ -0,0 +1,7 @@
+@import 'helpers'
+
+.op-modal--modal-body
+  max-height: 65vh
```
Is this something relevant for all modals? And why exactly `65vh`?

ingamai replied:
@HDinger It was kind of a 'the biggest height that fits most screen sizes' decision. But I removed this style now.

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial-form/ee-trial-form.component.sass:
```diff
@@ -0,0 +1,10 @@
+    .form--label::after
+      color: var(--content-form-danger-zone-bg-color)
+    .form--text-field
+      border: 1px solid var(--content-form-danger-zone-bg-color)
```
We already have a rule for the border, called `-required-highlighting`. So pease use this one. I guess it is sufficient to highlight the border, so this file can then be replaced completely.

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial-form/ee-trial-form.component.html:
```diff
@@ -0,0 +1,81 @@
+                 class="form--check-box"
+                 formControlName="newsletter_consent">
+        </div>
+        <span>I want to receive the OpenProject
```
Please localise the file.

HDinger commented on frontend/src/app/components/enterprise/enterprise-active-trial/ee-active-trial.component.html:
```diff
@@ -0,0 +1,38 @@
+      <div class="attributes-key-value--key">{{ text.label_maximum_users}}</div>
+      <div class="attributes-key-value--value-container">
+        <div class="attributes-key-value--value -text">
+          <span>10</span>
```
Is this value fixed? If yes, please move to the .ts file anyway.

HDinger commented on frontend/src/app/components/enterprise/enterprise-trial.service.ts:
```diff
@@ -0,0 +1,155 @@
+              readonly pathHelper:PathHelperService,
+              protected notificationsService:NotificationsService,
+              readonly formBuilder:FormBuilder) {
+    this.savedUserData = window.OpenProject.guardedLocalStorage(this.localStorageKey);
```
As discussed, we should think about a different solution here.

HDinger commented on frontend/src/app/components/enterprise/enterprise-base.component.html:
```diff
@@ -0,0 +1,17 @@
+
+<div *ngIf="eeTrialService.status === 'mailSubmitted'">
+  <enterprise-trial-waiting>
+    <enterprise-active-trial></enterprise-active-trial>
```
I assume that this should show the trial information within the modal? If yes, it doesn't work:
<img width="1097" alt="Bildschirmfoto 2020-03-12 um 13 21 23" src="https://user-images.githubusercontent.com/7457313/76521065-60b1e580-6464-11ea-84be-b46b17ad653e.png">

<img width="1088" alt="Bildschirmfoto 2020-03-12 um 13 19 53" src="https://user-images.githubusercontent.com/7457313/76521105-72938880-6464-11ea-8a13-eb93fdef00aa.png">

HDinger commented on frontend/src/app/components/enterprise/enterprise-trial.service.ts:
```diff
@@ -0,0 +1,155 @@
+
+  public confirmed:boolean;
+  public cancelled = false;
+  public status:string|undefined;
```
Please specify the types the `status` variable can be. So `public status:'startTrial'|'mailSubmitted'|undefined`.

HDinger reviewed (changes requested):
Hi @ingamai , looks good :+1: don't worry, most of the remarks are rather small :) 

Please merge the current dev, as there are merge conflicts and the tests cannot run. Some additional hints:

* When I submit the form and close the modal after that, there is a message saying that my Session has timed out. I think this is not the correct message here.
<img width="382" alt="Bildschirmfoto 2020-03-12 um 13 23 09" src="https://user-images.githubusercontent.com/7457313/76521344-f3eb1b00-6464-11ea-937b-26e51a03d7fd.png">

* Also when I close the modal before confirming my mail address, the page never updates after I confirmed. 

* When I press: "Try another email address", the form in the background is immediately changed. As a user I assume, that I've lost all my previous content and that I now cannot start the trial with that email (which is true, as the page never checks for the old mail again).

* The page looks a bit unstructured because there are many headlines, some have the same depth, some not. Maybe you could see, if there is room for improvement. However this is low priority.
<img width="1442" alt="Bildschirmfoto 2020-03-12 um 13 29 43" src="https://user-images.githubusercontent.com/7457313/76526268-8e4f5c80-646d-11ea-8362-a9d790020fd8.png">

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial.modal.ts:
```diff
@@ -0,0 +1,118 @@
+    // modal configuration
+    this.options = locals.options || {};
+    this.closeOnEscape = _.defaultTo(this.options.closeByEscape, false);
+    this.closeOnOutsideClick = _.defaultTo(this.options.closeByDocument, true);
```
Why do you set these variables so complicated? `locals.options` is not set when you call the modal, so all of this is somehow obsolete. You could also set `this.closeOnOutsideClick = true;`. The interface is then also not needed.

ingamai committed:
Review changes:
1. Localize all texts  and use 'email' consistently
2. checkmark added as html code
3. outsource iframe src (transform into trusted URL)
4. page headings restructured
5. modal configuration simplified
6. trial information also shown in modal window
7. codeclimate issues fixed

ingamai committed:
Save trial key in db to get user data from Augur

ingamai committed:
Fix: submit button enabled without general consent

ingamai committed:
Make form fields on mobile span the whole with

ingamai committed:
Fix merge conflicts while rebasing

ingamai commented:
> Hi @ingamai , looks good 👍 don't worry, most of the remarks are rather small :)
> 
> Please merge the current dev, as there are merge conflicts and the tests cannot run. Some additional hints:
> 
> * When I submit the form and close the modal after that, there is a message saying that my Session has timed out. I think this is not the correct message here.
> 
> * Also when I close the modal before confirming my mail address, the page never updates after I confirmed.
> * When I press: "Try another email address", the form in the background is immediately changed. As a user I assume, that I've lost all my previous content and that I now cannot start the trial with that email (which is true, as the page never checks for the old mail again).
> * The page looks a bit unstructured because there are many headlines, some have the same depth, some not. Maybe you could see, if there is room for improvement. However this is low priority.

@HDinger I rebased and hope I could fix all of your additional hints.

ingamai committed:
delete trial key from gon after mail was confirmed

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial-form/ee-trial-form.component.html:
```diff
@@ -0,0 +1,82 @@
+                 required>
+        </div>
+        <span>{{ text.general_consent }}
+          <a href="https://www.openproject.com/terms-of-service/" target="_blank">{{ text.terms_of_service }}</a>
```
Please move these links to the component as constants, too.

HDinger commented on frontend/src/app/components/enterprise/enterprise-modal/enterprise-trial.modal.html:
```diff
@@ -0,0 +1,73 @@
+                [disabled]="!eeTrialService.confirmed"
+                [hidden]="eeTrialService.status === 'startTrial'">
+        </button>
+        <button *ngIf="eeTrialService.status === 'startTrial'"
```
As said before, please extract all calls of this structure `eeTrialService.status === 'XY'` into methods.

HDinger commented on config/routes.rb:
```diff
@@ -340,6 +340,9 @@
     constraints(Enterprise) do
       resource :enterprise, only: %i[show create destroy]
+      scope controller: 'enterprises' do
+        post 'enterprise/create_trial_key' => 'enterprises#create_trial_key'
```
The name is misleading. The method is not really creating the trial key.

ingamai replied:
@HDinger I renamed it to "save_trial_key".

HDinger commented on frontend/src/app/components/enterprise/enterprise-active-trial/ee-active-trial.component.ts:
```diff
@@ -0,0 +1,74 @@
+        let savedData = this.eeTrialService.trialForm.value;
+        this.subscriber = savedData.first_name + ' ' + savedData.last_name;
+        this.email =  savedData.email;
+      }, 500);
```
Why this delay? It leads to a strange flickering effect, as you see the other form first, and then it is replaced.

ingamai replied:
@HDinger Moving the methods from the services into the component solved this issue as well :)

HDinger commented on frontend/src/app/components/enterprise/enterprise-trial.service.ts:
```diff
@@ -0,0 +1,208 @@
+
+  // use the trial key saved in the db
+  // to get the user data from Augur
+  private getUserDataFromAugur(trialKey:string) {
```
I have a little stomachache with these two method private methods:
First of all, why are they part of the service when they are actually only needed by the ActiveTrial component. This results in many unpleasant consequences: First, most of the data is not needed by the service (actually only the key). You can tell this by the fact that you simply write it into the trialForm, an object that will not be visible anymore. On the other hand, you will have to manually pull the data out of the trialForm in the component again. This is pretty ugly all in all. I think it would be better to drag the whole loading from gon into the component that needs it. Then you don't have to execute unnecessary code every time. You can then simply set the key from the component in the service.

HDinger commented on app/controllers/enterprises_controller.rb:
```diff
@@ -60,6 +65,20 @@ def destroy
+  end
+
+  def initialize_gon
+    @trial_key = Token::EnterpriseTrialToken.find_by!(user_id: current_user.id)
```
Please remove the `!`, it throws an exception when no token is found. Thus a user without a token has no chance to get one, as the page already throws an exception. 

<img width="816" alt="Bildschirmfoto 2020-03-26 um 14 42 44" src="https://user-images.githubusercontent.com/7457313/77660712-f62b9a00-6f79-11ea-80cb-9613e688ea6c.png">

Also I think that we should only look for tokens that have not expired yet, so something like `expires_on < Date.today`.

ingamai replied:
@HDinger Sorry, copy and paste failure... I removed it!

HDinger commented on app/views/enterprises/_current.html.erb:
```diff
@@ -1,43 +1,9 @@
-</div>
+<enterprise-active-trial data-subscriber="<%= @current_token.subscriber %>"
+                         data-email="<%= @current_token.mail %>"
+                         data-user-count="<%= @current_token.restrictions[:active_user_count] %>"
```
When I have an unlimited token, this causes the page to crash.
<img width="788" alt="Bildschirmfoto 2020-03-26 um 14 40 51" src="https://user-images.githubusercontent.com/7457313/77660944-44409d80-6f7a-11ea-8700-2bfa2eacd77c.png">

HDinger reviewed (changes requested):
Unfortunately there are still some things that are not quite yet complete. Here are the things I found. But please have a look at my comments from last time again. Some of them are still relevant.

* When I click through the modals to the onboarding video: The video reboots every few seconds because the polling is still running.
* In general, I had the impression several times that the polling was still running, even though it shouldn't. This also led to errors, but unfortunately I cannot reproduce them. But please make sure that you always stop polling correctly.
* As admin I can not delete the token created by another admin. This throws an error.
* If I have a ready token and enter it in the field at the bottom, the page will not change afterwards. The token will be saved correctly, but the page should also update. On the Rails console an error appears.
* When the time to confirm the mail has expired, a message appears that you should resend the mail. If I reload the page instead, I end up on the start page again and all my data is gone.
 
Rather unimportant:
* I don't think the name 'EnterpriseTrialToken' is well chosen. It implies that this is the real token, which it is not.
* If Admin1 and Admin2 request a token relatively simultaneously, the one who first confirmed the mail is accepted (in the example now Admin1). So far correct.  But if this is deleted, Admin2 suddenly sees the waiting screen again. I think both should see the start page again. The case is surely quite constructed, especially if you add the query that only tokens are loaded into gon that are not yet exposed.

ingamai commented:
@HDinger Maybe this is interesting for you: The reloading of the onboarding video had nothing to do with the polling but it seems to be a newer issue with binding the method that uses the DomSanitizer directly to the src attribute without a pipe.
Have a look at: https://stackoverflow.com/questions/39429293/url-sanitization-is-causing-refresh-of-the-embedded-youtube-video/48154761

ingamai committed:
Rename EnterpriseTrialToken and method in controller

ingamai committed:
Move  methods from service to components back again

ingamai committed:
Outsource youtube url and *ngIf conditions from html templates

ingamai committed:
Outsource all links into constants or static links, upsale icons added to images

HDinger committed:
Add EnterpriseBaseComponent to global dynamic components

ingamai committed:
Remove duplicate modal window padding

ingamai committed:
Fix pasting of a token and handle simultaneous changes by multiple admins

ingamai committed:
Fix Error msg after token has been saved

ingamai committed:
Fix user limit error for unlimited token, Fix error when deleting a pasted token

HDinger committed:
Extract status checks into methods and summarise strings

HDinger committed:
Code cleanup

oliverguenther committed:
Allow inserting a new token as previously possible

HDinger committed:
Adapt homescreen block to use same texts as the admin page and change the structure of the block

oliverguenther committed:
Add webmock chrome functionality with billy

HDinger committed:
Avoid double loading of components and remove unnecessary global variables

HDinger committed:
Handle cases where the user confirmed the mail after the browser stopped checking it. In this case the `/details` url returns an error. Without this additional check, the user is stuck for ever with the waiting screen.

HDinger committed:
Fix specs that broke due to the restructuring of the enterprise page

HDinger committed:
Replace images of Enterprise tease page

HDinger committed:
Match breadcrumb to page header

oliverguenther committed:
Test augur with puffing-billy to mock browser requests with a proxy

ingamai committed:
Display new user data if user updatet the data (tried with another email)

oliverguenther committed:
Add test case for reloading the page

HDinger committed:
Make used Augur API configurable

oliverguenther committed:
Use different components for active trial

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/31935/ee-activation

oliverguenther committed:
Fix blocks in chrome registration

ingamai committed:
Correct and add texts

ingamai committed:
Change page heading

ingamai committed:
Show other attributes in waiting screen

ingamai commented:
@oliverguenther I started working on this one https://community.openproject.com/projects/openproject/work_packages/32912/activity
by showing the missing attributes in the waiting screen.
I did not extend the structure of the tokens yet because this may be out of scope for 10.5. Do you think this is sufficient for now?

ulferts committed:
update Gemfile.lock after bundle run

RobinWagner commented:
Hi @ingamai,
to keep the discussion in one place I'll add my comments here instead of on the work package.

Thank you for making the adjustment to the strings.
The 4 benefits (Installation support, professional support, premium features, security features) look good and match with the strings that Niels specified in the PPT.

Unfortunately, I cannot see the strings used by the default screen (upper section) since there seems to be an error.
When I go to the Enterprise Edition screen (on PullPreview) there seems to be a token set for which "Subscriber: undefined undefined" is shown. Also, the status indicates that an email has already been sent (which is not the case). I tried this in multiple different browsers - also in incognito mode but it always shows this state. I'm not able to remove this default token. Also, requesting a separate token does not seem possible.

Could you take a look at this?
![image](https://user-images.githubusercontent.com/4333337/79232013-3db19180-7e67-11ea-81b2-e93ee8488779.png)

When I click on the link "Or try with another email address." and enter an email address an internal error is shown and the subscriber (while briefly changing to the provided name) quickly changes back to the original "undefined undefined".

![image](https://user-images.githubusercontent.com/4333337/79232148-6f2a5d00-7e67-11ea-9fdd-10c85493f22c.png)

When this is properly displayed again, I'm happy to take a look at the other strings.

Thank you,
Robin

oliverguenther committed:
Fix CSP for augur host

oliverguenther committed:
Use edge augur host as default for now

RobinWagner commented:
Hi @oliverguenther @ingamai,

Thank you for making the adjustment.
When I signed into PullPreview I still saw the default "undefined undefined" subscriber. But now it was possible to enter a different email address and receive the token. The internal error seems to be gone.

However, the displayed Enterprise Edition trial license seems to be off by 10 years. As a result, the trial license is immediately expired.
The end day is set to 2010 instead of 2020:
![image](https://user-images.githubusercontent.com/4333337/79319544-bf0d3080-7f08-11ea-9398-d01f8e4fdb12.png)

Inga and I already briefly spoke about this. @oliverguenther do you have an idea what could be causing this?

It would be great if we could make some small changes to the text. In some places some words seem to be missing. Also, there are two different spellings used for "Open Source". On our website we capitalized it and always wrote it as "Open Source". I would suggest we do this here as well:
![image](https://user-images.githubusercontent.com/4333337/79320022-64280900-7f09-11ea-8d3a-38506cc66348.png)

Maybe we should add a hint that we charge an extra (one-time) fee for the installation support. That may be more transparent. What do you think, @lindenthal?

When ordering the Enterprise Edition license, should it be activated automatically (like the trial license)?
When I tried to book / buy the Enterprise Edition I could fill out all the forms for the Enterprise Edition and go through the different steps (entering a test credit card number). However, after confirming nothing happened. The modal was hidden but the Enterprise Edition was not active. Instead, I received the Enterprise Edition by email.
Is that supposed to happen? Or should the Enterprise Edition license - similarly to the trial license be activated automatically?

Otherwise it looks good to me :)

Thank you,
Robin

oliverguenther committed:
Update strings

oliverguenther committed:
Skip trial spec for now

oliverguenther commented:
Hi @RobinWagner , thanks for your latest remarks. I'm going to go ahead and merge this for now, we can create subsequent tickets for your two questions if we decide to implement them
